### PR TITLE
simplify codec api

### PR DESCRIPTION
**Pull request by laizy**

### CODE REVIEW

File: ontio-std/src/abi/source.rs [hunk 1]
```diff
@@ -15,37 +18,57 @@ fn varuint_encode_size(val: u64) -> usize {
...
+    }
+
+    pub fn read<T: Decoder<'a>>(&mut self) -> Result<T, Error> {
+        T::decode(self)
+    }
+
+    pub(crate) fn read_address(&mut self) -> Result<&'a Address, Error> {
+        let buf = self.next_bytes(20)?;
+        Ok(unsafe { transmute(buf.as_ptr()) })
+    }
```
Comment (carltraveler) [inline ontio-std/src/abi/source.rs:53]: transmut will add trait vatable pointer to struct?

File: ontio-std/src/abi/source.rs [hunk 2]
```diff
@@ -15,37 +18,57 @@ fn varuint_encode_size(val: u64) -> usize {
...
+
+    pub(crate) fn read_h256(&mut self) -> Result<&'a H256, Error> {
+        let buf = self.next_bytes(32)?;
+        Ok(unsafe { transmute(buf.as_ptr()) })
+    }
```
Comment (carltraveler) [inline ontio-std/src/abi/source.rs:58]: why Result here state a 'a life time. can omit?

**Review by carltraveler — APPROVED** [review]

File: ontio-std/src/abi/source.rs (hunk 1 above)
Comment (laizy) [reply ontio-std/src/abi/source.rs:53]: no, just pointer cast.

File: ontio-std/src/abi/source.rs (hunk 2 above)
Comment (laizy) [reply ontio-std/src/abi/source.rs:58]: no. lifetime `'a` is longer than `&mut self`.

**Pull request merged**